### DISCUSSION TITLE
Grammar railroad diagram

### DISCUSSION THREAD

**Discussion by mingodad** [tree-sitter/tree-sitter#2269] [Libraries & tools]
I started a small attempt to generate an `EBNF` compatible with https://bottlecaps.de/rr/ui to be generated from the `grammar.json` you can see/play with it here https://mingodad.github.io/plgh/json2ebnf.html .

It already does a good job on most grammars but it can still be improved and any feedback is welcome.

You can choose any of the grammars listed in https://github.com/tree-sitter/tree-sitter/blob/master/docs/index.md through the `select` element or enter the `url` to the raw `grammar.json` to get it converted.

Cheers !

**Comment by mingodad** [comment]
The mentioned HTML page is self contained and also I have it here https://github.com/mingodad/plgh/blob/main/json2ebnf.html just in case someone want to contribute any pull request.

**Comment by ambroisie** [comment]
This is spam.

**Comment by mingodad** [comment]
Not really as I commented here https://github.com/mskelton/tree-sitter-mql5/issues/11#issuecomment-1549695778 .
I'm trying to make grammar writers aware of the utility of railroad diagrams.

**Comment by ambroisie** [comment]
You just sent the exact same comment to over 25 repositories.

I understand being enthusiastic about your project, but this is not good behaviour. Just open an issue with `tree-sitter` iteself (this very one) and talk about the feature here, rather that in 1001 different places.

**Comment by mingodad** [comment]
Thanks for your feedback !

**Comment by ahlinc** [comment]
@mingodad I saw your mentions about your `json2ebnf` tool many times in different places. It's awesome that such tool exist and maybe for someone it would be useful, but please don't do mass posting (there are enough spam on Github last times).
If you want to show something, please use discussions and don't open real issues, I'd like to minimize them in this repo and keep only important ones.

**Comment by mingodad** [comment]
I did an update to allow direct submit to https://www.bottlecaps.de/rr/ui and also added links to other parser related projects from the author (https://github.com/GuntherRademacher) of the railroad diagram generator.

**Comment by mingodad** [comment]
Hello @Harrypotter19876 !
I'm not sure what do you mean by `spammer account` .
Can you clarify that ?
Cheers !

**Comment by mingodad** [comment]
I'm trying to build an online yacc/lex (LALR(1)) grammar editor/tester to help develop/debug/document grammars the main repository is here https://github.com/mingodad/parsertl-playground and the online playground with several non trivial examples is here https://mingodad.github.io/parsertl-playground/playground/ .

Select a grammar/example from "Examples" select box and then click "Parse" to see a parser tree for the source in "Input source" editor.

It's based on https://github.com/BenHanson/gram_grep and https://github.com/BenHanson/lexertl14 .

Any feedback is welcome !

The grammars available so far (with varying state of correctness):

- Ada parser
- Akwa parser (not working)
- Aliceml parser (partially working)
- Ansi C11 parser (partially working)
- Ansi C18 parser (partially working)
- Ansi C parser
- Antlr4.5 parser
- Bison parser (partially working)
- Blawn parser
- Braille parser (not working)
- C3c  parser
- Calculator parser
- Carbon parser (need review of '*')
- Chapel parser
- CocoR parser (partially working)
- Cpp5-v2 parser (not working)
- Cql parser
- Cxx parser (not working)
- Doxygen code scanner torture
- Ecere parser (not working)
- GramGrep parser
- HTML parser
- idl2cpp parser
- Ispc parser
- Java11 parser
- JavascriptCore parser
- Jq parser (partially working)
- Json5 parser
- Json lexer
- Json parser
- LALR parser
- Lark parser
- Lezer parser (partially working)
- LFortran parser (partially working)
- Linden Script parser
- Lox parser
- LPegrex parser (partially working)
- LPython parser (not working)
- Lua2ljs parser
- Lua-5.3 parser
- Lua parser
- LuaPP parser (partially working)
- Minic parser
- Minizinc parser (not working)
- MSTA parser
- Mulang parser (not working)
- Openscad parser (partially working)
- Peg parser (partially working)
- Pikchr parser
- Playground3 parser
- Playground parser
- PnetC parser
- PnetCSHarp parser
- PnetDPas parser
- PnetJava parser
- PnetVBasic parser
- Postgresql parser (be patient)
- Preprocessor parser (not working)
- Rust parser
- Scheme parser
- Souffle parser
- Tameparse parser (not working)
- Textdiagram parser
- Textmapper parser
- Webassembly interpreter parser
- XML parser
- Z80 assembler parser

**Comment by mingodad** [comment]
I've just added `Tree-sitter-ebnf-generator` to https://mingodad.github.io/lua-wasm-playground/ (select `Tree-sitter-ebnf-generator` from `Examples` then click `Run` to see the converted grammar for the content in `Input Text (arg[1])`).

From https://github.com/eatkins/tree-sitter-ebnf-generator/tree/master :

Tree-sitter-ebnf-generator provides scripts that can convert files defined in an EBNF grammar into equivalent [tree-sitter](https://tree-sitter.github.io/tree-sitter/) grammar.js dsl format and vice versa. The EBNF syntax is both more compact and readable than the tree-sitter js syntax and far more similar to published EBNF grammars for languages such as scala and lua which should make writing grammars more straightforward and less error prone. The extended EBNF format defined in this project supports all of the features in the tree-sitter js DSL.

The examples directory shows how the ebnf format can be used to generate tree-sitter grammars for a language. The lua language context free grammar can be found [here](https://github.com/eatkins/tree-sitter-ebnf-generator/blob/master/examples/lua/lua.ebnf) along with its generated tree-sitter [grammar](https://github.com/eatkins/tree-sitter-ebnf-generator/blob/master/examples/lua/grammar.js). Similarly, scala language context free grammar can be found [here](https://github.com/eatkins/tree-sitter-ebnf-generator/blob/master/examples/scala/scala.ebnf) along with its generated tree-sitter [grammar](https://github.com/eatkins/tree-sitter-ebnf-generator/blob/master/examples/scala/grammar.js). All syntactical elements in the lua and scala grammars can be converted to the tree-sitter grammar.

**Reply by mingodad** [reply]
And enhanced it and added grammars converted with this `quickjs` script https://github.com/mingodad/plgh/blob/main/json2ebnf-lua.js .

Noticed that the best examples on how to use precedences are in Javascript/Typescript grammars, otherwise we get a grammar full of magic numbers.

List of grammars:
- Ada
- Bash
- cmake
- ComonLisp
- CPP
- CSS
- C
- D
- Dart
- Docker
- Elixir
- Elm
- Erlang
- Fortran
- Go
- Haskell
- Haxe
- HTML
- Javascript
- Java
- Julia
- Kotlin
- ld
- Lean
- Lua
- Make
- Markdown
- Minizinc
- Nim
- Ocaml
- PHP
- Python
- Ruby
- Rust
- Scala
- Typescript
- V
- VimScript
- Zig
